File: leetCode-75/day-6/(unfinished)102/589-n-array-tree-preorder-traversal.js
/**
 * Runtime: 74 ms
 * Memory Usage: 44.5 MB
 * Link: https://leetcode.com/problems/n-ary-tree-preorder-traversal/submissions/825362792/
 //-----------------------------------------------------------------------------
 //-----------------------------------------------------------------------------
 /*
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {

    // 1. Verify tgat tree is not empty.
    if(root == null) {
        return new Array();
    }

    // 2. Define a queue.
    var queue = new Queue();

    // 3. Enqueue the first element of the tree
    queue.enqueue(root);

    // 4. Define result array.
    var result = new Array();

    // 5. Loop while the queue containts elements.
    while(queue.size() > 0)
    {
        // 6. Define the iterative result array.
        var iterationResult = new Array();

        // 7. Pre save current queue size
        let count = queue.size();

        // 8. Loop over the queue elements that were in the queue at the begining of the iteration
        for(let i = 0; i < count; i++) {

            // 9. Dequeue element from the queue.
            let element = queue.dequeue();

            // 10. Push the element to the iteration result array.
            iterationResult.push(element.val)

            // 11. If the left child exist, add to the end ot the queue.
            if(element.left != null) {
                queue.enqueue(element.left)
            }

            // 12. If the right child exist, add to the end of the queue.
            if(element.right != null) {
                queue.enqueue(element.right)
            }
        }

        // 13. Add the iteration resut to the 
        result.push(iterationResult);
    }

    return result;
};

File: leetcode-75/day-6/589/589-n-array-tree-preorder-traversal.cs
namespace DefaultNamespace;

public class Solution {
    public IList<IList<int>> LevelOrder(TreeNode root) {
        
        if(root == null)
        {
            return new List<IList<int>>();
        }
        
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        var result = new List<IList<int>>();
            
        while(queue.Count > 0)
        {
            var count = queue.Count;
            var level = new List<int>();
            for(int i =0; i<count; i++)
            {
            
                var node = queue.Dequeue();
                level.Add(node.val);
            
                if(node.left != null)
                {
                    queue.Enqueue(node.left);
                }
                if(node.right != null)
                {
                    queue.Enqueue(node.right);
                }
            }
            
            result.Add(level);
        }
        
        return result;
    }
}
